TF-TRT Backward compatibility test

pooyadavoodi opened the pull request:
- Adds a new test to ensure TF-TRT can execute models saved in previous minor releases of TF2.x.
- The checked in model is generated using tensorflow-gpu==2.0.0

Commit:
Add script to generate TF-TRT model to be used for testing

Commit:
Add tftrt_2.0_saved_model

This model can be loaded and executed in tests to ensure
backwards compatibility across TF versions.

Commit:
Add backward compatibility test

Commit:
Move data of quantization_mnist_test to testdata/mnist

aaroey commented on tensorflow/python/compiler/tensorrt/test/quantization_mnist_test.py:
```diff
@@ -265,7 +265,7 @@ def _ModelFn(features, labels, mode):
     if not is_tensorrt_enabled():
       return
-    model_dir = test.test_src_dir_path('python/compiler/tensorrt/test/testdata')
+    model_dir = test.test_src_dir_path('python/compiler/tensorrt/test/testdata/mnist')
```
Where is the mnist test data?

pooyadavoodi replied:
ooops. Just added back.

aaroey commented on tensorflow/python/compiler/tensorrt/trt_convert_test.py:
```diff
@@ -837,6 +837,18 @@ def testTrtGraphConverter_StaticOp(self):
+      return
+
+    model_dir = test.test_src_dir_path(
+        'python/compiler/tensorrt/test/testdata/tftrt_2.0_saved_model')
```
This PR also checks in the variables of the test model, is that required even after the conversion?

pooyadavoodi replied:
I just added check for correctness which would need the variables.

Commit:
Add mnist dir, add correctness check for bwd compatibility test

Commit:
Fix pylint for backwards compatibility test

rthadur commented:
@pooyadavoodi can you please check build failures ?

Commit:
Sort targets in BUILD and shorten statement for pylint

pooyadavoodi commented:
I just fixed a few bugs.

Any idea about these errors which seem to be related to Eigen?

```
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/bitbucket.org/eigen/eigen/get/54bca9936424.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException GET returned 404 Not Found 
WARNING: Download from https://bitbucket.org/eigen/eigen/get/54bca9936424.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException Checksum was d8469c8be9ded39e914a7cd3c1246bae36ab0737d9c173328303766a3c7187e2 but wanted 9edd4860b52813eaf8c023f0de1767ec58e2d67a290b718e6702469208ac5be1
ERROR: An error occurred during the fetch of repository 'eigen_archive':
   java.io.IOException: Error downloading [https://storage.googleapis.com/mirror.tensorflow.org/bitbucket.org/eigen/eigen/get/54bca9936424.tar.gz, https://bitbucket.org/eigen/eigen/get/54bca9936424.tar.gz] to /tmpfs/bazel_output/_bazel_kbuilder/f2d52ca1f092ccbe254cc98c3dc90790/external/eigen_archive/54bca9936424.tar.gz: Checksum was d8469c8be9ded39e914a7cd3c1246bae36ab0737d9c173328303766a3c7187e2 but wanted 9edd4860b52813eaf8c023f0de1767ec58e2d67a290b718e6702469208ac5be1
Analyzing: 1471 targets (431 packages loaded, 3071 targets configured)
Analyzing: 1471 targets (440 packages loaded, 3250 targets configured)
ERROR: /tmpfs/src/github/tensorflow/third_party/eigen3/BUILD:33:1: //third_party/eigen3:eigen3 depends on @eigen_archive//:eigen in repository @eigen_archive which failed to fetch. no such package '@eigen_archive//': java.io.IOException: Error downloading [https://storage.googleapis.com/mirror.tensorflow.org/bitbucket.org/eigen/eigen/get/54bca9936424.tar.gz, https://bitbucket.org/eigen/eigen/get/54bca9936424.tar.gz] to /tmpfs/bazel_output/_bazel_kbuilder/f2d52ca1f092ccbe254cc98c3dc90790/external/eigen_archive/54bca9936424.tar.gz: Checksum was d8469c8be9ded39e914a7cd3c1246bae36ab0737d9c173328303766a3c7187e2 but wanted 9edd4860b52813eaf8c023f0de1767ec58e2d67a290b718e6702469208ac5be1
```

pooyadavoodi commented:
> I just fixed a few bugs.
> 
> Any idea about these errors which seem to be related to Eigen?
> 
> ```
> WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/bitbucket.org/eigen/eigen/get/54bca9936424.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException GET returned 404 Not Found 
> WARNING: Download from https://bitbucket.org/eigen/eigen/get/54bca9936424.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException Checksum was d8469c8be9ded39e914a7cd3c1246bae36ab0737d9c173328303766a3c7187e2 but wanted 9edd4860b52813eaf8c023f0de1767ec58e2d67a290b718e6702469208ac5be1
> ERROR: An error occurred during the fetch of repository 'eigen_archive':
>    java.io.IOException: Error downloading [https://storage.googleapis.com/mirror.tensorflow.org/bitbucket.org/eigen/eigen/get/54bca9936424.tar.gz, https://bitbucket.org/eigen/eigen/get/54bca9936424.tar.gz] to /tmpfs/bazel_output/_bazel_kbuilder/f2d52ca1f092ccbe254cc98c3dc90790/external/eigen_archive/54bca9936424.tar.gz: Checksum was d8469c8be9ded39e914a7cd3c1246bae36ab0737d9c173328303766a3c7187e2 but wanted 9edd4860b52813eaf8c023f0de1767ec58e2d67a290b718e6702469208ac5be1
> Analyzing: 1471 targets (431 packages loaded, 3071 targets configured)
> Analyzing: 1471 targets (440 packages loaded, 3250 targets configured)
> ERROR: /tmpfs/src/github/tensorflow/third_party/eigen3/BUILD:33:1: //third_party/eigen3:eigen3 depends on @eigen_archive//:eigen in repository @eigen_archive which failed to fetch. no such package '@eigen_archive//': java.io.IOException: Error downloading [https://storage.googleapis.com/mirror.tensorflow.org/bitbucket.org/eigen/eigen/get/54bca9936424.tar.gz, https://bitbucket.org/eigen/eigen/get/54bca9936424.tar.gz] to /tmpfs/bazel_output/_bazel_kbuilder/f2d52ca1f092ccbe254cc98c3dc90790/external/eigen_archive/54bca9936424.tar.gz: Checksum was d8469c8be9ded39e914a7cd3c1246bae36ab0737d9c173328303766a3c7187e2 but wanted 9edd4860b52813eaf8c023f0de1767ec58e2d67a290b718e6702469208ac5be1
> ```

That error is hopefully fixed by this: https://github.com/tensorflow/tensorflow/commit/e3a7bdbebb99352351a19e2e403136166aa52934

aaroey commented:
Also @sanjoy

aaroey commented on tensorflow/python/compiler/tensorrt/test/testdata/gen_tftrt_model.py:
```diff
@@ -0,0 +1,129 @@
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
```
There is an error when trying to submit this PR:
'import tensorflow' statements are no allowed inside TF's code base... Please import modules directly.

Also could you help to add a build rule for this file?

pooyadavoodi replied:
This file is not supposed to be used by any test. It's supposed to be used manually by developers to generate a test model if needed. Do we still need to add it to BUILD?

aaroey replied:
I think adding a build rule will make our life easier: we just need to run `bazel run` to get the model. But I don't have strong opinion. But at least replace `import tensorflow` with direct module imports.

pooyadavoodi replied:
I removed the import.

I also noticed the test model was missing in the data dependency of trt_convert_test which is fixed now. I missed this because I run tests without bazel :(

I haven't added gen_tftrt_model.py to BUILD. I would need to learn how to do that first. I suggest to postpone this for now.

Commit:
Remove import tensorflow from gen_tftrt_model.py

Commit:
Add tftrt_2.0_saved_model to data deps of trt_convert_test in BUILD

Also remove gen_tftrt_model.py from BUILD because this file is
not supposed to be used by any test. It's supposed to be
used manually by developers to generate a test model if needed.

aaroey reviewed (dismissed):
Thanks for the fix Pooya!

pooyadavoodi commented:
Getting this now:
```
Missing the following dependencies from pip_packages:

Missing dependency: //tensorflow/python/compiler/tensorrt:test/testdata/tftrt_2.0_saved_model/variables/variables.index
Affected Tests:
//tensorflow/python/compiler/tensorrt:test/testdata/tftrt_2.0_saved_model/variables/variables.index (null)
//tensorflow/python/compiler/tensorrt:trt_convert_test (70635f4b39bbe1d08d95efe3e379ef53)

Missing dependency: //tensorflow/python/compiler/tensorrt:test/testdata/tftrt_2.0_saved_model/variables/variables.data-00000-of-00002
Affected Tests:
//tensorflow/python/compiler/tensorrt:test/testdata/tftrt_2.0_saved_model/variables/variables.data-00000-of-00002 (null)
//tensorflow/python/compiler/tensorrt:trt_convert_test (70635f4b39bbe1d08d95efe3e379ef53)

Missing dependency: //tensorflow/python/compiler/tensorrt:test/testdata/tftrt_2.0_saved_model/variables/variables.data-00001-of-00002
Affected Tests:
//tensorflow/python/compiler/tensorrt:test/testdata/tftrt_2.0_saved_model/variables/variables.data-00001-of-00002 (null)
//tensorflow/python/compiler/tensorrt:trt_convert_test (70635f4b39bbe1d08d95efe3e379ef53)

Missing dependency: //tensorflow/python/compiler/tensorrt:test/testdata/tftrt_2.0_saved_model/saved_model.pb
Affected Tests:
//tensorflow/python/compiler/tensorrt:test/testdata/tftrt_2.0_saved_model/saved_model.pb (null)
//tensorflow/python/compiler/tensorrt:trt_convert_test (70635f4b39bbe1d08d95efe3e379ef53)
Traceback (most recent call last):
  File "pip_smoke_test.py", line 184, in <module>
    main()
  File "pip_smoke_test.py", line 177, in main
    or add no_pip tag to the test.""")
RuntimeError:
    One or more added test dependencies are not in the pip package.
If these test dependencies need to be in TensorFlow pip package, please add them to //tensorflow/tools/pip_package/BUILD.
Else either blacklist the dependencies in //tensorflow/tools/pip_package/pip_smoke_test.py
or add no_pip tag to the test.
```

What do you suggest?

aaroey committed:
Add no_pip tag to avoid pip_smoke_test depending on newly added savedmodel.

aaroey committed:
Merge branch 'master' into tftrt_bwd_compatibility_test

aaroey reviewed (changes requested):
Unapprove and try to trigger the copybara import again.
